pkg/cloudprovider/provider/cherryservers/provider.go

```diff
@@ -0,0 +1,333 @@
+	"github.com/kubermatic/machine-controller/pkg/providerconfig"
+	"golang.org/x/crypto/ssh"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/common"
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+)
+
+const privateRSAKeyBitSize = 4096
+
+type provider struct {
+	configVarResolver *providerconfig.ConfigVarResolver
+}
+
+type RawConfig struct {
+	TeamID    providerconfig.ConfigVarString `json:"teamID"`
+	ProjectID providerconfig.ConfigVarString `json:"projectID"`
+	Token     providerconfig.ConfigVarString `json:"token,omitempty"`
+	Plan      providerconfig.ConfigVarString `json:"plan"`
+	Location  providerconfig.ConfigVarString `json:"location"`
+	Labels    map[string]string              `json:"labels,omitempty"`
+}
+
+type Config struct {
+	TeamID    string
+	ProjectID string
+	Token     string
+	Plan      string
+	Location  string
+	Labels    map[string]string
+}
+
+func New(configVarResolver *providerconfig.ConfigVarResolver) cloudprovidertypes.Provider {
+	return &provider{configVarResolver: configVarResolver}
+}
+
+func (p *provider) AddDefaults(spec v1alpha1.MachineSpec) (v1alpha1.MachineSpec, error) {
+	return spec, nil
+}
+
+func (p *provider) getConfig(s v1alpha1.ProviderSpec) (*Config, *providerconfig.Config, error) {
+	if s.Value == nil {
+		return nil, nil, fmt.Errorf("machine.spec.providerconfig.value is nil")
+	}
+	pconfig := providerconfig.Config{}
+	err := json.Unmarshal(s.Value.Raw, &pconfig)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	rawConfig := RawConfig{}
+	if err = json.Unmarshal(pconfig.CloudProviderSpec.Raw, &rawConfig); err != nil {
+		return nil, nil, err
+	}
+
+	c := Config{}
+	c.Token, err = p.configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.Token, "CS_TOKEN")
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"token\" field, error = %v", err)
+	}
+	c.Plan, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Plan)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	c.ProjectID, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.ProjectID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"projectID\" field, error = %v", err)
+	}
+
+	c.TeamID, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.TeamID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"teamID\" field, error = %v", err)
+	}
+
+	c.Location, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Location)
+	if err != nil {
+		return nil, nil, err
+	}
+	c.Labels = rawConfig.Labels
+	return &c, &pconfig, err
+}
+
+func getNameForOS(os providerconfig.OperatingSystem) (string, error) {
+	switch os {
+	case providerconfig.OperatingSystemUbuntu:
+		return "Ubuntu 18.04 64bit", nil
+	case providerconfig.OperatingSystemCentOS:
+		return "CentOS 7 64bit", nil
+	}
+	return "", providerconfig.ErrOSNotSupported
+}
+
+func getClient(token string) *cherrygo.Client {
+	os.Setenv("CHERRY_AUTH_TOKEN", token)
```

Comment:
In the documentation, you mention `CHERRYSERVERS_TOKEN`. Which one is supposed to be used?

Reply:
cherrygo client API needs CHERRY_AUTH_TOKEN to be set. I jave created CHERRYSERVERS_TOKEN to be inline with other providers. Then I make the provider set CHERRY_AUTH_TOKEN to the value of CHERRYSERVERS_TOKEN taken from manifest

Reply:
Fixed